src/colorfire.1
.TH "OGLC Screensaver" 1 "" "X Version 11"
.de Ds
.Sp
.nf
..
.de De
.fi
..
.SH NAME
colorfire - Color-fire-explosion-thing-effect.
.SH SYNOPSIS
.B colorfire
[\--root/-r]
[\--maxfps/-x \fInumber\fP]
[\--vsync/-y \fInumber\fP]
[\--dpms/-M \fInumber\fP]
[\--texture/-t \fInumber\fP]
.SH DESCRIPTION
From Andreas Gustafsson (\fIhttp://www.lysator.liu.se/~sdw/opengl2000.html\fP):
"This didn't take me long to do, and I didn't write more than approximately 100 lines of code, but the result is still not that bad!"

Ported to Linux by Tugrul Galatali.
.SH OPTIONS
.TP 8
.B \--root
Draw on the root window.
.TP 8
.B \--maxfps \fInumber\fP
Set maximum frame rate.
.TP 8
.B \--vsync \fInumber\fP
Limit redraws to specified number of vertical refreshes.  0 - 100.  Default: 1
.TP 8
.B \--dpms \fInumber\fP
Stop rendering new frames if the display is not on.  0 - 1.  Default: 1
.TP 8
.B \--texture \fInumber\fP
.Ds
0 - None
1 - Smoke (--smoke)
2 - Ripples (--ripples)
3 - Ripples (--smooth)
.De

Default: Random
.SH ENVIRONMENT
.PP
.TP 8
.B DISPLAY
to get the default host and display number.
.SH SEE ALSO
.BR X (1),
.SH COPYRIGHT
Copyright \(co 1999 by Andreas Gustafsson and Tugrul Galatali.  

Colorfire is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License version 2 as
published by the Free Software Foundation.

Colorfire is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
.SH AUTHOR
Andreas Gustafsson <sdw@lysator.liu.se> and Tugrul Galatali <tugrul@galatali.com>



ChangeLog
******************************* Version 0.9.1  *********************************

	Added port of Pixel City, originally by Shamus Young,
http://code.google.com/p/pixelcity/
	Replaced methods deprecated by ImageMagick.

******************************* Version 0.9.0  *********************************

	Added port of drempels, originally by Ryan M. Geiss,
http://www.geisswerks.com/about_drempels.html
	Added new saver feedback by Tugrul Galatali.
	Added support for DPMS to idle the saver when the display isn't on.
	Fixed brown paper bag bug in hyperspace.
	Added two new presets to flocks inspired by Matt Richard's Wavy Gravy on
openprocessing.org
	Tweaked hufo_smoke to avoid flat flame effect.
	Fixed unknown argument segfault in lattice.
	Switched matrixview's main character array to vertex arrays, which helps
significantly on older nVidia hardware.
	hyperspace and drempels have command line arguments to disable use of
shaders in case they behave unexpectedly.
	Minor changes to fix builds on other platforms.

******************************* Version 0.8.3 *********************************

	New screen saver Lorenz contributed by Soeren Sonnenburg <sonne@debian.org>
	Fix for SF #2673397 by changing =='s to >='s in widthChange update logic,
otherwise non-sense values for interpolation would wreak havoc. Also changed for
xyzChange.

******************************* Version 0.8.2 *********************************

	Limit frame rate to vertical refresh by default if compiled with GLEW
and GLX_SGI_swap_control is available. Otherwise, maxfps set to 30.
	--nice eliminated and made default. nanosleep/usleep are much better 
behaved these days.
	spirographx modified to use vertex arrays and GNU extension sincosf if
available, dropping CPU usage significantly.
	Minor updates to correct issues noted by newer compilers.
	--disable-debug is now the default in configure.

******************************* Version 0.8.1 *********************************

	Pulled in vroot.h patch from Ari Pollak's Debian package.
	Added missing #include <assert.h> to hyperspace.cpp
	Image loading in matrixview was hanging for whatever reason,
		rewritten with the new Magick Wand interface.
	ImageMagick now required.
	timeval to double conversion now fixed with an explicit cast of tv_sec
		to double.
	Fixed automake files to use CLEANFILES
	Dropped SSE/3dnow stuff
	Fixed hyperspace.xml typo in dist_kdessconfig_DATA

******************************* Version 0.8.0 *********************************

	Rough sync against new Windows release, rssavers 0.1
		Except for skyrocket
		Added new saver Hyperspace

******************************* Version 0.7.6 *********************************

	Fix to matrixview.c where the body of a thread was accidentally left as a
subfunction of a function that exits while the thread is still alive.

	Added an autoconf check for ExportImagePixels since it was very recently
added (5.5.7).

******************************* Version 0.7.5 *********************************

	Removed multiline strings that are no longer supported by gcc 3.3.

	Added spirographx, a port of an OSX app/screensaver SpirographX by 
Holmes Futrell <holmes@neatosoftware.com>.
	http://homepage.mac.com/solidmag/spiro/

	Added matrixview, based on Alex Zolotov's Matrix GL. Hacked in external 
image loading and ability to modify the resolution of the matrix.

******************************* Version 0.7.4 *********************************

	Fixed bug in lattice when the variable passed to acos would be slightly
greater than 1, causing cascading nan's that made the program useless.

	Explicitly clearing buffers to avoid garbage in blurring modes.

	Lots of stupid little bug fixes.

******************************* Version 0.7.3 *********************************

	Added trails and blur options to flocks. Also added presets to spruce
things up in combination with random preset selection when running 
unconfigured.

	Fixed another cpu detection bug, damn off by one errors :)

	Added support for decimal integer arguments to -window-id. 

	Added .desktop files for preliminary support of KDE. Not sure where
the default directory is, but if it isn't specified by --with-kdessconfigdir,
the configure script checks for the existence of:

/usr/share/applnk/System/ScreenSavers

******************************* Version 0.7.2 *********************************

	Fixed bug in cpu detection code.

	Fixed a bug that included audio data in skyrocket when OpenAL wasn't 
found.

******************************* Version 0.7.1 *********************************

	Fixed a typo crashing skyrocket on big endian machines. Spotted by
Michel Danzer.

	Two new savers: biof and busyspheres.

	New texture smooth for colorfire.

	Fixed a bug that left screensavers running after they should have quit.

	Replaced compile time cpu feature detection with runtime detection.

******************************* Version 0.7.0 *********************************

	Extracted all binary data from the source into binary files. The
source distribution has gotten notably smaller as a result. 

	The binary files are now converted during compilation to significantly 
larger source files with escaping instead of base64 type inefficient strings.
The resulting executables have no unnecessary bloat.

	If the bzip2 library is detected on the system, it is used to compress
the binary data before conversion.

	Added --disable-bzip2 flag to configure. This is to give up space in
order to avoid a delay while decompressing, which is noticeable in skyrocket
that has about 3MB of compressed textures and audio.

	Added --disable-sound flag to configure. This chops off about 3.3MB
(before compression) of audio from skyrocket and any runtime overhead
associated with sound. Useful if you intend to keep passing volume = 0 to
skyrocket.

	Added SSE/3DNow optimized code that gives euphoria a nice 10% boost
on my P4, probably a bit more on an Athlon but I haven't had recent console 

src/colorfire.cpp
/*
 * Copyright (C) 1999  Andreas Gustafsson
 * Ported to Linux by Tugrul Galatali <tugrul@galatali.com>
 *
 * Colorfire is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as 
 * published by the Free Software Foundation.
 *
 * Colorfire is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

// Colorfire screen saver

#include <math.h>
#include <stdio.h>
#include <GL/gl.h>
#include <GL/glu.h>

#include "colorfire_textures.h"
#include "driver.h"
#include "loadTexture.h"
#include "rsRand.h"

const char *hack_name = "Colorfire by Andreas Gustafsson (Shadow/Noice) (C)1999";

int dTexture;

#define NR_WAVES 24
#define TEXSIZE 256

float wrot[NR_WAVES], wtime[NR_WAVES], wr[NR_WAVES], wg[NR_WAVES], wb[NR_WAVES], wspd[NR_WAVES], wmax[NR_WAVES];
float v1 = 0, v2 = 0, v3 = 0;

void initWave (int nr)
{
	wtime[nr] = 0;
	wrot[nr] = rsRandf (360);
	wr[nr] = rsRandf (1.0);
	wg[nr] = rsRandf (1.0);
	wb[nr] = rsRandf (1.0);
	wspd[nr] = 0.5f + rsRandf (0.3f);
	wmax[nr] = 1.0f + rsRandf (1.0f);
}

void hack_reshape (xstuff_t * XStuff)
{
	// Window initialization
	glViewport (0, 0, (GLsizei) XStuff->windowWidth, (GLsizei) XStuff->windowHeight);
	glMatrixMode (GL_PROJECTION);
	glLoadIdentity ();
	gluPerspective (55.0, (GLfloat) XStuff->windowWidth / (GLfloat) XStuff->windowHeight, 1.0, 20.0);
	glMatrixMode (GL_MODELVIEW);
	glLoadIdentity ();
	glTranslatef (0.0, 0.0, -3.0);
	glClearColor (0.0, 0.0, 0.0, 0.0);
}

void hack_init (xstuff_t * XStuff)	// Called right after the window is created, and OpenGL is initialized.
{
	int i;

	hack_reshape (XStuff);

	for (i = 0; i < NR_WAVES; i++) {
		initWave (i);
		wtime[i] = 3.0f + rsRandf (1.0f);
	}

	if (dTexture) {
		unsigned char *l_tex;
		unsigned int tex;

		glGenTextures (1, &tex);

		glBindTexture (GL_TEXTURE_2D, tex);

		glTexParameterf (GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_REPEAT);
		glTexParameterf (GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_REPEAT);
		glTexParameterf (GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
		glTexParameterf (GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR_MIPMAP_NEAREST);

		switch (dTexture) {
		case 1:
			LOAD_TEXTURE (l_tex, smokemap, smokemap_compressedsize, smokemap_size);
			gluBuild2DMipmaps (GL_TEXTURE_2D, 1, TEXSIZE, TEXSIZE, GL_RGB, GL_UNSIGNED_BYTE, l_tex);
			FREE_TEXTURE (l_tex)

			break;
	
		case 2:
			LOAD_TEXTURE (l_tex, ripplemap, ripplemap_compressedsize, ripplemap_size);
			gluBuild2DMipmaps (GL_TEXTURE_2D, 1, TEXSIZE, TEXSIZE, GL_RGB, GL_UNSIGNED_BYTE, l_tex);
			FREE_TEXTURE (l_tex)

			break;

		case 3:
			int i, j;
			double x, y, r, d;
			unsigned char texbuf[64][64];

			memset ((void *)&texbuf, 0, 4096);

			r = 32;
			for (i = 0; i < 64; i++) {
				for (j = 0; j < 64; j++) {
					x = abs (i - 32);
					y = abs (j - 32);
					d = sqrt (x * x + y * y);

					if (d < r) { 
						d = 1 - (d / r);
						texbuf[i][j] = char (255 * d * d);
					}
				} 
			}

			gluBuild2DMipmaps (GL_TEXTURE_2D, 1, 64, 64, GL_LUMINANCE, GL_UNSIGNED_BYTE, texbuf);

			glTexParameteri (GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR_MIPMAP_NEAREST);
			glTexParameteri (GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);

			break;
		}
	} else {
		unsigned char img[] = { 0, 255, 255, 0 };

		gluBuild2DMipmaps (GL_TEXTURE_2D, 1, 2, 2, GL_LUMINANCE, GL_UNSIGNED_BYTE, img);
		glTexImage2D (GL_TEXTURE_2D, 0, 1, 2, 2, 0, GL_LUMINANCE, GL_UNSIGNED_BYTE, img);
	}

	glEnable (GL_TEXTURE_2D);
	glBlendFunc (GL_SRC_ALPHA, GL_ONE);
	glEnable (GL_BLEND);
}

void hack_cleanup (xstuff_t * XStuff)
{
}

void drawWave (int nr, float fDeltaTime)
{
	float colMod = 1.0;

	glPushMatrix ();
	glScalef (wtime[nr], wtime[nr], wtime[nr]);
	if (wtime[nr] > (wmax[nr] - 1.0))
		colMod = wmax[nr] - wtime[nr];
	glRotatef (wrot[nr], 1.f, 0, 0);
	glRotatef (wrot[nr], 0, 1.f, 0);
	glRotatef (wrot[nr], 0, 0, 1.f);
	glColor3f (colMod * wr[nr], colMod * wg[nr], colMod * wb[nr]);
	glBegin (GL_QUADS);
	glTexCoord2f (0, 0);
	glVertex3f (-1, -1, 0);
	glTexCoord2f (1, 0);
	glVertex3f (1, -1, 0);
	glTexCoord2f (1, 1);
	glVertex3f (1, 1, 0);
	glTexCoord2f (0, 1);
	glVertex3f (-1, 1, 0);
	glEnd ();
	wtime[nr] += fDeltaTime * wspd[nr];
	if (wtime[nr] > wmax[nr])
		initWave (nr);
	glPopMatrix ();
}

void hack_draw (xstuff_t * XStuff, double currentTime, float frameTime)
{
	glClear (GL_COLOR_BUFFER_BIT);
	glPushMatrix ();
	glRotatef (v1, 1.f, 0, 0);
	glRotatef (v2, 0, 1.f, 0);
	glRotatef (v3, 0, 0, 1.f);
	v1 += frameTime * 5;
	if (v1 > 360)
		v1 -= 360;
	v2 += frameTime * 10;
	if (v2 > 360)
		v2 -= 360;
	v3 += frameTime * 7;
	if (v3 > 360)
		v3 -= 360;
	for (int i = 0; i < NR_WAVES; i++)
		drawWave (i, frameTime);
	glPopMatrix ();
}

void hack_handle_opts (int argc, char **argv)
{
	int change_flag = 0;

	dTexture = 1;

	while (1) {
		int c;

#ifdef HAVE_GETOPT_H
		static struct option long_options[] = {
			{"help", 0, 0, 'h'},
			DRIVER_OPTIONS_LONG
			{"texture", 1, 0, 't'},
			{"smoke", 0, 0, 1},
			{"ripples", 0, 0, 2},
			{"smooth", 0, 0, 3},
			{0, 0, 0, 0}
		};

		c = getopt_long (argc, argv, DRIVER_OPTIONS_SHORT "ht:", long_options, NULL);
#else
		c = getopt (argc, argv, DRIVER_OPTIONS_SHORT "ht:");
#endif
		if (c == -1)
			break;

		switch (c) {
			DRIVER_OPTIONS_CASES 
		case 'h':printf ("%s:"
#ifndef HAVE_GETOPT_H
							      " Not built with GNU getopt.h, long options *NOT* enabled."
#endif
							      "\n" DRIVER_OPTIONS_HELP "\t--texture/-t <arg>\n" "\t--smoke\n" "\t--ripples\n" "\t--smooth\n", argv[0]);

			exit (1);
		case 't':
			change_flag = 1;
			dTexture = strtol_minmaxdef (optarg, 10, 0, 3, 0, 1, "--texture: ");
			break;
		case 1:
		case 2:
		case 3:
			change_flag = 1;
			dTexture = c;
			break;
		}
	}

	if (!change_flag) {
		dTexture = rsRandi (3) + 1;
	}
}
